reactor-extra/src/main/java/reactor/cache/CacheHelper.java

```diff
@@ -45,54 +46,141 @@
 public class CacheHelper {
 
 	/**
-	 * Restore a {@link Mono<Value>} from the cache-map given a provided key. If no value
+	 * Restore a {@link Mono Mono&lt;VALUE&gt;} from the cache-map given a provided key. If no value
 	 * is in the cache, it will be calculated from the original source which is set up in
 	 * the next step. Note that if the source completes empty, this result will be cached
-	 * and all subsequent requests with the same key will return {@link Mono#empty()). The
+	 * and all subsequent requests with the same key will return {@link Mono#empty()}. The
 	 * behaviour is similar for erroring sources, except cache hits would then return
 	 * {@link Mono#error(Throwable)}.
+	 * <p>
+	 * Note that the wrapped {@link Mono} is lazy, meaning that subscribing twice in a row
+	 * to the returned {@link Mono} on an empty cache will trigger a cache miss then a
+	 * cache hit.
 	 *
-	 * @param cache {@link Map} wrapper of a cache
+	 * @param cm {@link Map} wrapper of a cache
 	 * @param key mapped key
-	 * @param <Key> Key Type
-	 * @param <Value> Value Type
+	 * @param <KEY> Key Type
+	 * @param <VALUE> Value Type
 	 *
 	 * @return Lazy "{@link MonoCacheBuilderMapMiss}"
 	 */
-	@NonNull
-	public static <Key, Value> MonoCacheBuilderMapMiss<Value> lookupMono(@NonNull Map<Key, ? super Signal<? extends Value>> cm,
-			@NonNull Key key) {
-		return o -> Mono.justOrEmpty(cm.get(key))
-		                .switchIfEmpty(o.materialize()
-		                                .doOnNext(value -> cm.put(key, value)))
-		                .dematerialize();
+	public static <KEY, VALUE> MonoCacheBuilderMapMiss<VALUE> lookupMono(Map<KEY, ? super Signal<? extends VALUE>> cm, KEY key) {
```

Comment:
Does it make sense to defer it in this way? Since cached value is already cold and it makes sense to provide supplier here 
```
MonoCacheBuilderCacheWriter<KEY, VALUE> onCacheMissResume(Supplier<Mono<? extends VALUE>> other);
```

Reply:
Also agreed with @smaldini comment, it would be clearer if we have separate helpers for flux and mono :)

Reply:
it is more an issue of being able to reuse the `Mono` returned by the last method in the chain, and still get cache hits:

```java
Mono<Integer> cached = CacheHelper.lookupMono(map, key)
    .onCacheMissResume(Mono.just(1));

cached.subscribe(); //cache miss & load
assertThat(map).containsKey(key);
cached.subscribe(); // STILL GOES THROUGH CACHE MISS PATH
```

Reply:
Aha! Got it! Yup, my bad, missed that... But what about a supplier for cacheMiss? Is it needed at all?

Reply:
it could make sense, added it